bam_qc_metrics/bam_qc.py

```diff
@@ -403,6 +364,99 @@ def validate_config_fields(self, config):
             # do not log this message; logger not yet initialized
             raise ValueError(msg)
 
+    def write_downsampled_bam(self, bam_path, sample_level, total_reads):
+        """
+        Write a temporary downsampled BAM file:
+        - bam_path is the input BAM file
+        - sample_level is number of reads desired in the downsampled file
+        - total_reads is number of reads in the file denoted by bam_path
+
+        Uses `samtools -s`
+        Argument to `samtools -s` is of the form RANDOM_SEED.DECIMAL_RATE
+        Eg. for random seed 42 and sample rate of 1 in 4, 42 + (1/4) = 42.25
+        DECIMAL_RATE = (SAMPLE_LEVEL * MULTIPLIER) / TOTAL_READS
+
+        Number of reads output by `samtools -s` is only approximate
+        see https://github.com/samtools/samtools/issues/931
+        Use a multiplier > 1 to ensure sufficient number of reads; then cut down the output
+        """
+        downsampled_path = None
+        msg = "Starting downsampling with sample level %i, total reads %i" \
+              % (sample_level, total_reads)
+        self.logger.info(msg)
+        multiplier = 1.2
+        if self.random_seed == self.DEFAULT_RANDOM_SEED:
+            self.logger.info("Using default random seed %i", self.DEFAULT_RANDOM_SEED)
+        else:
+            self.logger.info("Using custom random seed %i", self.random_seed)
+        sample_decimal = float(sample_level*multiplier) / total_reads
+        sample_arg = str(self.random_seed + sample_decimal)
+        self.logger.debug("Sampling argument to 'samtools view' is %s", sample_arg)
+        ds_raw_path = os.path.join(self.tmpdir, 'downsampled.bam')
+        pysam.view('-u', '-s', sample_arg, '-o', ds_raw_path, bam_path, catch_stdout=False)
+        sampled = int(pysam.view('-c', ds_raw_path).strip())
+        self.logger.debug("%i reads in initial downsampled file" % sampled)
+        if sampled < sample_level:
+            # shouldn't happen with appropriate multiplier, but check for completeness
+            msg = "Downsampling reads: Expected at least %i, found %i" % (sample_level, sampled)
+            self.logger.error(msg)
+            raise ValueError(msg)
+        elif sampled == sample_level:
+            # shouldn't happen with appropriate multiplier, but check for completeness
+            msg = "Found %i downsampled reads; no further correction needed" % sample_level
+            self.logger.debug(msg)
+            downsampled_path = ds_raw_path
+        else:
+            # sampled reads > desired total, need to cut down
+            # choose reads to remove at random (with known seed) from the initial sampled set
```

Comment:
I don't think this takes into account paired reads. Paired reads should be treated as one unit.

Reply:
Just read your metrics document. I think it's better to be around approximately 1 million reads rather than lose random reads from read pairs.

Reply:
No, it doesn't. That occurred to me after I made the PR. I don't think this will affect many metrics other than `PairsMappedAbnormallyFar`. For now I have updated the documentation: https://github.com/oicr-gsi/bam-qc-metrics/pull/17/commits/6a31e2eab7f26b5db62e287c03930e190dde83f7 

For a proper solution, we could use `samtools collate` as suggested by the samtools developers, at the cost of running a process in a subshell. Alternatively we could try and implement the samtools collate command in Python.

Thoughts?

Reply:
Re. the above error -- good point. sample_decimal should be the computed value or 1, whichever is less.